src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3060,21 +3060,25 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
               setError(fun)
             }
           } else {
-            fun.body match {
+            if (pt.typeSymbol == PartialFunctionClass) {
               // translate `x => x match { <cases> }` : PartialFunction to
               // `new PartialFunction { def applyOrElse(x, default) = x match { <cases> } def isDefinedAt(x) = ... }`
-              case Match(sel, cases) if (sel ne EmptyTree) && (pt.typeSymbol == PartialFunctionClass) =>
-                // go to outer context -- must discard the context that was created for the Function since we're discarding the function
-                // thus, its symbol, which serves as the current context.owner, is not the right owner
-                // you won't know you're using the wrong owner until lambda lift crashes (unless you know better than to use the wrong owner)
-                val outerTyper = newTyper(context.outer)
-                val p = vparams.head
-                if (p.tpt.tpe == null) p.tpt setType outerTyper.typedType(p.tpt).tpe
-
-                outerTyper.synthesizePartialFunction(p.name, p.pos, paramSynthetic = false, fun.body, mode, pt)
-
-              case _ => doTypedFunction(fun, resProto)
-            }
+              val tree = fun.body match {
+                case Match(sel, cases) if sel ne EmptyTree => fun.body
+                case _                                     =>
```

Comment:
Also, I find it clearer to bind the default case, rather than referring back to the scrutinee.

```suggestion
                case funBody                               =>
```